### PR TITLE
fix(pageserver): drain upload queue before offloading timeline

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

It is possible (is it?) at the point we shutdown the timeline, there are still layer files we did not upload.

## Summary of changes

* If the queue is not empty, avoid offloading.
  * Hmm... is it correct? We don't explicitly wait for files to be uploaded in the compaction code, and we offload the timeline right after compaction, which means that it will always fail if a compaction gets scheduled and triggers timeline offload?
* Otherwise, shutdown the timeline gracefully using the flush mode to ensure all local files are uploaded before deleting the timeline directory.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/timeline/offload.rs [hunk 1]
```diff
@@ -33,6 +34,21 @@ pub(crate) async fn offload_timeline(
     debug_assert_current_span_has_tenant_and_timeline_id();
     tracing::info!("offloading archived timeline");
 
+    if !timeline.remote_client.no_pending_work() {
+        return Err(OffloadError::RemoteStorage(anyhow!(
+            "cannot offload timeline with pending uploads"
+        )));
+    }
+
+    // Drain again in case of race condition (i.e., some background tasks add new uploads right after we checked)
```
Comment (arpad-m) [inline pageserver/src/tenant/timeline/offload.rs:43]: this doesn't prevent race conditions, we could have a background upload right after this.

File: pageserver/src/tenant/timeline/offload.rs [hunk 2]
```diff
@@ -33,6 +34,21 @@ pub(crate) async fn offload_timeline(
     debug_assert_current_span_has_tenant_and_timeline_id();
     tracing::info!("offloading archived timeline");
 
+    if !timeline.remote_client.no_pending_work() {
+        return Err(OffloadError::RemoteStorage(anyhow!(
+            "cannot offload timeline with pending uploads"
```
Comment (arpad-m) [inline pageserver/src/tenant/timeline/offload.rs:39]: I think this should be a separate `OffloadError` variant, that we silently ignore in compaction (or log an info line).

File: pageserver/src/tenant/timeline/offload.rs (hunk 1 above)
Comment (skyzh) [reply pageserver/src/tenant/timeline/offload.rs:43]: Yes, so I also set the upload mode to be `FreezeAndFlush` to ensure everything gets eventually uploaded. Doing a wait-for-completion early can reduce the shutdown time.

Comment (arpad-m) [reply pageserver/src/tenant/timeline/offload.rs:43]: In that case the comment is wrong, but I also think it is not really needed: shutdown shouldn't take that long in the first place.

**Review by arpad-m — APPROVED** [review]

**Pull request merged**